### PR TITLE
Fixes #12208 - Seed script for katello-foreman migration

### PR DESCRIPTION
**Pull request by parthaa**

### CODE REVIEW

File: db/seeds.rb [hunk 1]
```diff
@@ -4,158 +4,9 @@
...
-]
-
-notifications.each do |notification|
-  ::MailNotification.find_or_create_by_name(notification)
+katello_seeds = Dir.glob("#{Katello::Engine.root}/db/seeds.d/*.rb"
+katello_seeds.each do |seed|
+  puts "Seeding #{seed}" unless Rails.env.test?
+  load seed
 end
+puts "All katello seed files executed" unless Rails.env.test?
```
Comment (ehelms) [inline db/seeds.rb]: This file shouldn't be needed anymore -- foreman/config/seeds.rb should handle loading all of these

File: db/seeds.d/02-seeds.rb [hunk 2]
```diff
@@ -0,0 +1,161 @@
+# This file should contain all the record creation needed to seed the database with its default values.
```
Comment (ehelms) [inline db/seeds.d/02-seeds.rb]: If we are breaking seeds up, can we break up the rest of this file into logical seed files?

File: db/migrate/20140707203534_location_add_katello_default.rb [hunk 3]
```diff
@@ -1,6 +1,6 @@
 class LocationAddKatelloDefault < ActiveRecord::Migration
   def up
-    add_column :taxonomies, :katello_default, :boolean, :null => false, :default => true
+    add_column :taxonomies, :katello_default, :boolean, :null => false, :default => false
```
Comment (ehelms) [inline db/migrate/20140707203534_location_add_katello_default.rb]: Why do we want to change an existing migration like this?

File: db/seeds.d/01-locations.rb [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# This file should contain all the record creation needed to seed the database with its default values.
```
Comment (ehelms) [inline db/seeds.d/01-locations.rb]: Maybe name this 01-default-location.rb ?

File: db/migrate/20140707203534_location_add_katello_default.rb (hunk 3 above)
Comment (parthaa) [inline db/migrate/20140707203534_location_add_katello_default.rb]: We need to change this for the following reason. 'katello_default' would automatically get set to true for all locations that were in foreman before this migrate script ran. We need to have only one katello_default

Comment (ehelms) [inline db/migrate/20140707203534_location_add_katello_default.rb]: Makes sense - thanks for the explanation.

**Comment by ehelms** [comment]
Since you had mentioned adding tests, going to set this in WoC state until you say otherwise.

**Comment by ehelms** [comment]
Looks like this needs a rebase

File: db/seeds.d/102-organizations.rb [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  User.current = nil
+end
+
+if ENV['SEED_ORGANIZATION']
+  Organization.without_auditing do
+    User.current = User.anonymous_admin
+    org = Organization.find_by_name!(ENV['SEED_ORGANIZATION'])
+    ForemanTasks.sync_task(::Actions::Katello::Organization::Create, org) unless org.library
+    User.current = nil
+  end
```
Comment (ehelms) [inline db/seeds.d/102-organizations.rb]: Won't the condition starting on line 7 handle this case given that the organization would have been created in an earlier seed ?

File: db/seeds.d/104-proxy.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+#
+# !!! PLEASE KEEP THIS SCRIPT IDEMPOTENT !!!
+#
+
+def format_errors(model = nil)
+  return '(nil found)' if model.nil?
+  model.errors.full_messages.join(';')
+end
+
+::User.current = ::User.anonymous_api_admin
```
Comment (ehelms) [inline db/seeds.d/104-proxy.rb]: Does this need to be set back to nil at the end?

File: db/seeds.d/105-roles-permissions.rb [hunk 7]
```diff
@@ -0,0 +1,69 @@
+# This file should contain all the record creation needed to seed the database with its default values.
+# The data can then be loaded with the rake db:seed (or created alongside the db with db:setup).
+#
+# !!! PLEASE KEEP THIS SCRIPT IDEMPOTENT !!!
+#
+User.current = ::User.anonymous_api_admin
```
Comment (ehelms) [inline db/seeds.d/105-roles-permissions.rb]: Does this need to be unset at the end of the seed?

File: db/seeds.d/106-mail-notifications.rb [hunk 8]
```diff
@@ -0,0 +1,34 @@
+# This file should contain all the record creation needed to seed the database with its default values.
+# The data can then be loaded with the rake db:seed (or created alongside the db with db:setup).
+#
+# !!! PLEASE KEEP THIS SCRIPT IDEMPOTENT !!!
+#
+::User.current = ::User.anonymous_api_admin
```
Comment (ehelms) [inline db/seeds.d/106-mail-notifications.rb]: Does this need to be unset at the end?

File: db/seeds.d/107-enable-dynflow.rb [hunk 9]
```diff
@@ -0,0 +1,9 @@
+# This file should contain all the record creation needed to seed the database with its default values.
+# The data can then be loaded with the rake db:seed (or created alongside the db with db:setup).
+#
+# !!! PLEASE KEEP THIS SCRIPT IDEMPOTENT !!!
+#
+
+::User.current = ::User.anonymous_api_admin
```
Comment (ehelms) [inline db/seeds.d/107-enable-dynflow.rb]: Same question

File: db/seeds.d/108-ensure-sync-notification.rb [hunk 10]
```diff
@@ -0,0 +1,10 @@
+# This file should contain all the record creation needed to seed the database with its default values.
+# The data can then be loaded with the rake db:seed (or created alongside the db with db:setup).
+#
+# !!! PLEASE KEEP THIS SCRIPT IDEMPOTENT !!!
+#
+::User.current = ::User.anonymous_api_admin
```
Comment (ehelms) [inline db/seeds.d/108-ensure-sync-notification.rb]: Same question

File: test/lib/tasks/seeds_test.rb [hunk 11]
```diff
@@ -0,0 +1,45 @@
+require 'test_helper'
+
+class SeedsTest < ActiveSupport::TestCase
+  # Disable AR transactional fixtures as we use DatabaseCleaner's truncation
+  # to empty the DB of fixtures for testing the seed script
```
Comment (ehelms) [inline test/lib/tasks/seeds_test.rb]: Should this comment be above line 8 instead of here?

File: test/lib/tasks/seeds_test.rb [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+    assert Location.default_location.present?
+    assert_equal "seed_test", Location.default_location.name
+  end
+
+  test "setup org bindings for every org in foreman" do
+    org = Organization.create!(:name => "myOrg")
+    ForemanTasks.expects(:sync_task).with(::Actions::Katello::Organization::Create, org)
+    seed
+  end
+end
```
Comment (ehelms) [inline test/lib/tasks/seeds_test.rb]: Might as well add tests for the other seeds to ensure provisioning templates are configured properly, permissions are added, mail notifications are set up etc.

Comment (parthaa) [inline test/lib/tasks/seeds_test.rb]: Added

File: db/seeds.d/102-organizations.rb [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+
+if ENV['SEED_ORGANIZATION']
+  if Setting['db_pending_seed']
+    admin = User.where(:login => ENV['SEED_ADMIN_USER'].present? ? ENV['SEED_ADMIN_USER'] : 'admin').first
+    if admin && admin.default_organization.nil?
+      admin.default_organization = Organization.find_by_name!(ENV['SEED_ORGANIZATION'])
+      admin.save!
+    end
+  end
+end
```
Comment (ehelms) [inline db/seeds.d/102-organizations.rb]: Do you recall why we do this assignment here? I am blanking.

File: db/seeds.d/103-provisioning-templates.rb [hunk 14]
```diff
@@ -0,0 +1,37 @@
+# This file should contain all the record creation needed to seed the database with its default values.
+# The data can then be loaded with the rake db:seed (or created alongside the db with db:setup).
+#
+# !!! PLEASE KEEP THIS SCRIPT IDEMPOTENT !!!
+#
+
+prior = ::User.current
```
Comment (ehelms) [inline db/seeds.d/103-provisioning-templates.rb]: Why is this needed?

**Comment by ehelms** [comment]
The seed task declaration can now be removed from engine.rb (https://github.com/Katello/katello/blob/master/lib/katello/engine.rb#L193-L195)

**Comment by daviddavis** [comment]
Just out of curiosity, why are the ruby files in `db/seeds.d` hyphenated?

**Comment by ehelms** [comment]
@daviddavis following convention https://github.com/theforeman/foreman/tree/develop/db/seeds.d

File: app/models/katello/concerns/location_extensions.rb [hunk 15]
```diff
@@ -8,6 +8,7 @@ module LocationExtensions
         before_create :set_katello_default
         before_save :refute_katello_default_changed
         before_destroy :deletable?
+        attr_accessible :katello_default
```
Comment (daviddavis) [inline app/models/katello/concerns/location_extensions.rb]: Do we want people to be able to set/update this field in the API? If not, I'd recommend removing this and using the following in our seed file(s):

```ruby
Location.create!(:name => "Default Location") do |location|
  location.katello_default = true
end
```

**Comment by daviddavis** [comment]
@ehelms ah ok. Although it doesn't look like this quite follows the convention (e.g. `103-provisioning-templates.rb` vs `07-provisioning_templates.rb`)

**Comment by ehelms** [comment]
@daviddavis touche, @parthaa see https://github.com/Katello/katello/pull/5539#issuecomment-162057238

File: app/models/katello/concerns/location_extensions.rb (hunk 15 above)
Comment (ehelms) [inline app/models/katello/concerns/location_extensions.rb]: Probably not, :+1:

**Comment by parthaa** [comment]
[test]

File: app/models/katello/concerns/location_extensions.rb (hunk 15 above)
Comment (parthaa) [inline app/models/katello/concerns/location_extensions.rb]: Believe me I tried hard not to add this. Its there because 
```
Location.create!(:name => "Default Location") do |location|
  location.katello_default = true
end
```
Didnt work. I would get "cant mass assign" even though its technically not mass assigning there,

File: db/seeds.d/102-organizations.rb (hunk 13 above)
Comment (parthaa) [inline db/seeds.d/102-organizations.rb]: I am just following what the seed originally had there.

File: app/models/katello/concerns/location_extensions.rb (hunk 15 above)
Comment (parthaa) [inline app/models/katello/concerns/location_extensions.rb]: Look at the output when I tried exactly your changes http://ci.theforeman.org/job/test_katello_pr_core/957/database=postgresql,ruby=1.9.3/console

**Comment by parthaa** [comment]
@ehelms @daviddavis success!!

**Comment by daviddavis** [comment]
ACK from me as long as we don't have `attr_accessible :katello_default` in location extensions.

File: app/models/katello/concerns/location_extensions.rb (hunk 15 above)
Comment (daviddavis) [inline app/models/katello/concerns/location_extensions.rb]: Did you try this too?

```ruby
location = Location.new(:name => 'default')
location.katello_default = true
location.save!
```

Comment (daviddavis) [inline app/models/katello/concerns/location_extensions.rb]: Actually, it looks like the code did work. In the job you linked to, the db seed completely successfully. However, it looks like it fails on this line: https://github.com/Katello/katello/pull/5539/files#diff-816d256f98117950a20b7ae031224817R71.

**Comment by daviddavis** [comment]
Looks good to me. @ehelms can you review too?

**Comment by ehelms** [comment]
ACK
